File: 0126-word-ladder-ii/0126-word-ladder-ii.rb
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {String[][]}
def backtrack(paths, begin_word, parents)
  return paths.select{ |path| path[0] == begin_word } if paths[0][0] == begin_word
  ans = []
  paths.each{ |path| parents[path[0]].each{ |parent| ans << [parent]+path } }
  backtrack(ans, begin_word, parents)
end

def find_ladders(begin_word, end_word, word_list)
  layers, visited, word_set, parents = [[begin_word].to_set], [begin_word].to_set, word_list.to_set, Hash.new{ |h, k| h[k] = Set.new }
  
  while !visited.include?(end_word) && !layers[-1].empty?
    curr_layer = layers[-1]
    layers << Set.new
    curr_layer.each do |curr_word|
      (0...curr_word.length).each do |j|
        ('a'..'z').each do |c|
          next_word = curr_word[0...j]+c+curr_word[j+1..-1]
          if next_word != curr_word && word_set.include?(next_word) && !visited.include?(next_word)
            next_layer = layers[-1]
            next_layer << next_word
            parents[next_word] << curr_word
          end
        end
      end
    end
    visited |= layers[-1]
  end
  
  return [] if !visited.include?(end_word)
  backtrack([[end_word]], begin_word, parents)
end

File: 0126-word-ladder-ii/0126-word-ladder-ii.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {string[][]}
 */
var findLadders = function(beginWord, endWord, wordList) {
  const layers = [new Set([beginWord])];
  const visited = new Set([beginWord]);
  const wordSet = new Set(wordList);
  const parents = {};
  
  while (!visited.has(endWord) && layers.at(-1).size) {
    const currLayer = layers.at(-1);
    layers.push(new Set());
    currLayer.forEach(currWord => {
      for (let j = 0; j < currWord.length; j++) {
        for (let k = 0; k < 26; k++) {
          const nextWord = currWord.slice(0, j) + String.fromCharCode('a'.charCodeAt(0) + k) + currWord.slice(j + 1);
          if (nextWord !== currWord && wordSet.has(nextWord) && !visited.has(nextWord)) {
            const nextLayer = layers.at(-1);
            nextLayer.add(nextWord);
            parents[nextWord] = parents[nextWord] || new Set();
            parents[nextWord].add(currWord);
          }
        }
      }
    });
    layers.at(-1).forEach(word => visited.add(word));
  }
  if (!visited.has(endWord)) return [];
  
  const backtrack = paths => {
    if (paths[0][0] === beginWord) return paths.filter(path => path[0] === beginWord);
    const ans = [];
    paths.forEach(path => parents[path[0]].forEach(parent => ans.push([parent, ...path])));
    return backtrack(ans);
  };
  
  return backtrack([[endWord]]);
};